### DISCUSSION TITLE
Multiple Prisma connections and schemas

### DISCUSSION THREAD

**Discussion by rdrigos** [prisma/prisma#24247] [Q&A] [answered]
### Question

Good morning, I have a project that I am developing that needs to deal with multiple databases, initially two, both Sql Server:

Database 1: Server=database1;Database=tesla1;User Id=sa;Password=mysecret;TrustServerCertificate=true;
Database 2: Server=database2;Database=tesla2;User Id=sa;Password=mysecret;TrustServerCertificate=true;

and needed to create schemas for both to create their tables. How could I do that? So when I do:

``` npm prism migrate ```

Create the tables in their respective database.

Would a command like this work?
npx prisma migrate dev --schema=prisma/schema1.prisma --datasource=db1
npx prisma migrate dev --schema=prisma/schema2.prisma --datasource=db2

**Comment by ludralph** [comment, marked as answer]
Hi @rdrigos 👋 

To work with multiple databases, you can create separate schema files for each database. In each schema file, you define the datasource with the respective database connection string. Here's an example of how you can set up your schema files:
For Database 1, create a schema file `prisma/db-1/schema.prisma`:
```prisma
datasource db {
  provider = "sqlserver"
  url      = env("DATABASE_URL_DB_1")
}

generator client {
  provider        = "prisma-client-js"
  output          = "../../node_modules/@prisma-db-1/client"
}
```
For Database 2, create a schema file `prisma/db-2/schema.prisma`:
```prisma
datasource db {
  provider = "sqlserver"
  url      = env("DATABASE_URL_DB_2")
}

generator client {
  provider        = "prisma-client-js"
  output          = "../../node_modules/@prisma-db-2/client"
}
```
In your .env file, you should define the `DATABASE_URL_DB_1` and `DATABASE_URL_DB_2` environment variables with your respective database connection strings.
To generate clients and run migrations, you need to specify the schema:
```
npx prisma generate --schema prisma/db-1/schema.prisma
npx prisma generate --schema prisma/db-2/schema.prisma

npx prisma migrate dev --schema prisma/db-1/schema.prisma --name initial
npx prisma migrate dev --schema prisma/db-2/schema.prisma --name initial
```
In your code, you can import the generated clients like this:
```ts
import { PrismaClient as PrismaClient1} from '@prisma-db-1/client';
import { PrismaClient as PrismaClient2} from '@prisma-db-2/client';
```
This way, `PrismaClient1` connects to Database 1 and `PrismaClient2` connects to Database 2.

Currently, Prisma does not support running a single migrate command for multiple schemas. You need to run the migrate command separately for each schema.

**Reply by rdrigos** [reply]
Ok, see if I understood correctly, I will need to create Prisma Schemas for each database:

```
root
   |---- prisma
               |---- rbsdb
               |         |---- schema.prisma (Configured with the respective output and URL)
               |---- x3v12
               |          |---- schema.prisma (Configured with the respective output and URL)
```

And when generating migrations, I use the commands:
```npx prisma migrate dev --schema prisma/rbsdb/schema.prisma --name initial```
```npx prisma migrate dev --schema prisma/x3v12/schema.prisma --name initial```

and use it like this:
```import { PrismaClient as PrismaClient1} from '@prisma-db-1/client';```
```import { PrismaClient as PrismaClient2} from '@prisma-db-2/client';```

or create a file ```src/libs/prisma.ts```:
```
import { PrismaClient as PrismaClient1} from '@prisma-db-1/client';
import { PrismaClient as PrismaClient2} from '@prisma-db-2/client';

const Client1 = new PrismaClient1();
const Client2 = new PrismaClient2();

export {
     Client1,
     Client2
}
```

**Reply by ludralph** [reply]
Yes, that's correct.

**Reply by rdrigos** [reply]
Okay, I'll test this later, and I'll report back with the results.
Thanks

**Reply by simibruda** [reply]
I have done like this but is not generated tve Prisma.Where types

**Reply by rdrigos** [reply]
The types are generated in the folder where the output was configured, in this example:

```
node_modules/@prisma-db-1/client
```

**Reply by simibruda** [reply]
Yes but I generated the clients but i have the error that the where and include types anre not generated

**Reply by rdrigos** [reply]
I never received this error, so I don't have the necessary knowledge to help you! I'm sorry!

I'm marking the @ludralph, because I see that he helps people in the Prisma community a lot and I believe he has the necessary knowledge to answer you

**Reply by simibruda** [reply]
It will be great if @ludralph could help me with this issue

**Reply by ludralph** [reply]
Hi @simibruda, Can you please open a new discussion and fill out the template with the necessary information to help debug this issue?

**Reply by simibruda** [reply]
sure https://github.com/prisma/prisma/discussions/24870

**Reply by IArnaut2** [reply]
Have you heard of DRY principle?

Programmers should NOT repeat themselves by making schemas and repeating models. They should be able to create just ONE schema for multiple databases.

**Comment by anuoua** [comment]
I think Prisma should consider splitting schema datasource and generator. It must be more flexible, can do a lot of things like multiple connections.
